kep/399-sysched-scoring/README.md

```diff
@@ -0,0 +1,161 @@
+ 
+### Goals
+
+1. No changes to the core scheduler; need only implement the features as a Scoring plugin
+2.  No impact when the scoring plugin is disabled
+3.  When scoring plugin is enabled:
+	-  Works in conjunction with other scoring plugins
+	-  No impact if pod does not have an associated system call profile
+	-  Improve pod security:
+
+4.  Reduce the number of victim pods that can be impacted by a "bad" pod
+5.  Reduce the number of nodes in a cluster that can be impacted by a "bad" pod
+6.  Minimizes the "excess" system calls across all pods on a node and all nodes in a cluster
+7.  Minimizes attack surface areas of a node (reduce needed system calls on a node)
+
+### Non-goals
+
+1.  Use other pod security attributes as part of scheduling metric (only considering system call in initial version)
+2.  Use existing Taint mechanism to implement above
+3.  Descheduling or rescheduling a pod after placement if the system call condition on the node changes during a pod's lifetime.
+4.  Automatically generating a system call profile of a pod
+
+## Proposal
+
+### User Stories
+
+Administrator deploys the scheduler with the SySched scoring plugin enabled. Users may deploy their pod as usual. As pods are scheduled, the pods are automatically placed according to the pod's specified constraints and our security metric discussed above.
+
+ #### Story 1
+
+<p align="center">
+<img src="images/example_complete.png" style="width:80%">
+</p>
+
+<p align = "center">
+<b>Figure 1. Scheduling algorithm and pod placement example</b>
+</p>
+
+Figure 1  illustrates how our scoring algorithm works and how pods are placed. The figure shows the placement of six pods (C1, C2, C3, C4, C5,and  C6)  in  two  nodes  (Node 1  and  Node 2).  The  numbers inside a rectangular box for a pod indicate the system calls the pod use, i.e., C1 {1, 2, 3, 5}, C2 {4, 7, 8}, C3 {1, 3, 5, 9}, C4 {2, 5, 6, 8}, C5 {2, 4, 5, 7}, and C6 {2, 7, 8}. Initially, to schedule pod C1, the ExS  scores for both Node 1 and Node 2 are the same (i.e., 0), since there are no pods running on the nodes (Figure 1(a)). Thus, our algorithm randomly picks one of the two nodes, Node 1 in this case, and updates its system call usage list using the system calls of C1 (Figure 1(b)). The ExS  scores for scheduling pods C2 are 7 and 0 for Node 1 and Node 2, respectively (Figure 1(b)). In this case, the scheduling algorithm picks Node 2 since Node 2 has a lower ExS  score. Figure 1(c) shows the updated system call lists for Node 2. This process repeats for all incoming pods. After all pods are placed, the cluster-wide ExS score in this example is 12, i.e., ExS1 is 5 and ExS2 is 7, where 1 and 2 are node numbers.
```

Comment:
Can you be more specific about the meaning of the numbers (in C1 {1, 2, 3, 5})? Like it's a particular type of system call?